File: app/src/main/java/com/softartdev/notecrypt/ui/note/NotePresenter.java
package com.softartdev.notecrypt.ui.note;

import android.text.TextUtils;

import com.softartdev.notecrypt.App;
import com.softartdev.notecrypt.db.DbStore;
import com.softartdev.notecrypt.model.Note;

import javax.inject.Inject;

public class NotePresenter {
    private NoteView mView;
    private Note mNote;

    @Inject
    DbStore dbStore;

    NotePresenter(NoteView view) {
        App.getAppComponent().inject(this);
        mView = view;
    }

    void createNote() {
        mNote = dbStore.createNote();
    }

    void saveNote(final String title, final String text) {
        if (mNote == null) {
            createNote();
        }

        if (TextUtils.isEmpty(title) && TextUtils.isEmpty(text)) {
            mView.onEmptyNote();
        } else {
            dbStore.saveNote(mNote.getId(), title, text);
            mView.onSaveNote(title);
        }
    }

    void loadNote(final long noteId) {
        mNote = dbStore.loadNote(noteId);
        mView.onLoadNote(mNote.getTitle(), mNote.getText());
    }

    void deleteNote() {
        if (mNote != null) {
            dbStore.deleteNote(mNote.getId());
        }
        mNote = null;
        mView.onDeleteNote();
        mView.onNavBack();
    }

    void checkSaveChange(String title, String text) {
        String savedTitle = mNote.getTitle();
        String savedText = mNote.getText();
        if (!title.equals(savedTitle) || !text.equals(savedText)) {
            mView.onCheckSaveChange();
            return;
        }
        if (savedTitle.equals("") && savedText.equals("")) {
            deleteNote();
        }
        mView.onNavBack();
    }
}

File: app/src/main/java/com/softartdev/notecrypt/ui/note/NotePresenter.kt
package com.softartdev.notecrypt.ui.note

import android.text.TextUtils
import com.softartdev.notecrypt.data.DataManager
import com.softartdev.notecrypt.di.ConfigPersistent
import com.softartdev.notecrypt.model.Note
import com.softartdev.notecrypt.ui.base.BasePresenter
import timber.log.Timber
import javax.inject.Inject

@ConfigPersistent
class NotePresenter @Inject
constructor(private val dataManager: DataManager) : BasePresenter<NoteView>() {
    private var mNote: Note? = null

    fun createNote() {
        addDisposable(dataManager.createNote()
                .subscribe({note ->
                    mNote = note
                }, {it.printStackTrace() }))
    }

    fun loadNote(noteId: Long) {
        addDisposable(dataManager.loadNote(noteId)
                .subscribe({note ->
                    mNote = note
                    mvpView?.onLoadNote(note.title, note.text)
                }, { it.printStackTrace() }))
    }

    fun saveNote(title: String, text: String) {
        if (mNote == null) {
            createNote()
        }

        if (TextUtils.isEmpty(title) && TextUtils.isEmpty(text)) {
            mvpView?.onEmptyNote()
        } else {
            addDisposable(dataManager.saveNote(mNote!!.id, title, text)
                    .subscribe(
                            { mvpView?.onSaveNote(title) },
                            { it.printStackTrace() }))
        }
    }

    fun deleteNote() {
        addDisposable(dataManager.deleteNote(mNote!!.id)
                .subscribe(
                        { Timber.d("Note deleted") }
                        , { it.printStackTrace() }))
        mNote = null
        mvpView?.onDeleteNote()
        mvpView?.onNavBack()
    }

    fun checkSaveChange(title: String, text: String) {
        val changed = dataManager.checkChanges(mNote!!.id, title, text)
        if (changed) {
            mvpView?.onCheckSaveChange()
            return
        }
        val empty = dataManager.emptyNote(mNote!!.id)
        if (empty) {
            deleteNote()
        }
        mvpView?.onNavBack()
    }
}
